jaeger: add check sub command

Pothulapati opened the pull request:
This adds a new `linkerd jaeger check` command to have checks w.r.t
jaeger extension. This is similar to that of the `linkerd check` cmd.
As jaeger is a separate package, It was a bit complex for this to work
as not all types and fields from healthcheck pkg are public, Helper
funcs were used to mitigate this.

This has the following changes:

- Adds a new `check.go` file under the jaeger extension pkg
- Moves some commonly needed funcs and types from `cli/cmd/check.go`
  and `pkg/healthcheck/health.go` into
  `pkg/healthcheck/healthcheck_output.go`.

Example Output
```bash
tarun in dev in on  kind-kind () linkerd2 on  tarun/jaeger-check [$!?] via 🐹 v1.15.4 took 2s
❯ ./bin/go-run cli jaeger check
linkerd-jaeger
--------------
√ collector service account exists
√ jaeger service account exists
√ collector config map exists
‼ collector pod is running
    collector-69cc44dfbc-8qkxw status is Pending
‼ jaeger pod is running
    jaeger-6f98d5c979-5vnpt status is Pending
√ jaeger extension pods are injected

Status check results are √

tarun in dev in on  kind-kind () linkerd2 on  tarun/jaeger-check [$!?] via 🐹 v1.15.4 took 2s
❯ ./bin/go-run cli jaeger check
linkerd-jaeger
--------------
√ collector service account exists
√ jaeger service account exists
√ collector config map exists
√ collector pod is running
√ jaeger pod is running
√ jaeger extension pods are injected

Status check results are √
```

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

adleong commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2399,17 +2445,20 @@ func getPodStatuses(pods []corev1.Pod) map[string]map[string][]corev1.ContainerS
 
 	for _, pod := range pods {
-		if pod.Status.Phase == corev1.PodRunning && strings.HasPrefix(pod.Name, "linkerd-") {
+		if pod.Status.Phase == corev1.PodRunning {
```
This existing logic here seems very confusing and overly complex and seems to have a lot of assumptions about the naming scheme of the Linkerd control plane baked in.  I think we should probably just rewrite these helper functions to be simpler.  But instead of doing that here, can we just leave this stuff alone for now and create more straightforward methods to use for the jaeger checks?  

All we need is a helper for determining if a container is running in a set of pods.  That should be as simple as iterating through the pods and their container statuses until we find a container with a matching name and then checking its status.

Pothulapati replied:
makes sense!

Pothulapati replied:
Why not check for the pod Status itself? 🤔 Could be simpler. I did that rn. WDYT?

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,189 @@
+func (options *checkOptions) checkFlagSet() *pflag.FlagSet {
+	flags := pflag.NewFlagSet("check", pflag.ExitOnError)
+
+	flags.StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace to use for --proxy checks (default: all namespaces)")
```
This should be used for data plane checks, but I don't see those implemented here.

Pothulapati replied:
Removed it.

Should we not have a flag for user to be able to specify which ns to check for the jaeger extension pods checks? 🤔 Currently, Its dynamic (ie based on labels) as you suggested!

adleong reviewed (commented):
I don't see checks that the linkerd-proxy sidecar is present in for this extensions components or a check to see if data plane pods have the correct env variables injected.  Will those checks be follow ups to this PR?

Pothulapati commented:
@adleong Yep, Plan to make the dataplane checks in a separate PR!

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,257 @@
+		proxyContainer := false
+		for _, containerSpec := range pod.Spec.Containers {
+			if containerSpec.Name == k8s.ProxyContainerName {
+				proxyContainer = true
```
can't we just immediately return nil here?

Pothulapati replied:
We can do that but this way we can also report which pod does not have the proxy container :thinking: but its limited only for the first pod!

adleong replied:
oh, I see, this is inside the pods loop.  I think this would be clearer if we had a method that takes a pod and checks if that pod has a proxy container.  If we did it that way then we could immediately return nil from that method when a proxy contain is detected.  With that in place, this method becomes much simpler since we can immediately return an error when we encounter a pod without a proxy container.

Pothulapati replied:
But we would be still needing the boolean flag to check for the proxy container right? I updated the fun to be different ones, as you suggested!

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,257 @@
+// checkPodsStatus checks if the pod is in running state
+func checkPodsStatus(pods []corev1.Pod) error {
+	for _, pod := range pods {
+		fmt.Println(pod.Name, pod.Status.Phase)
```
is this leftover debugging output?

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,257 @@
+					return err
+				}
+
+				pods, err := kubeAPI.GetPodsByNamespace(ctx, jaegerNamespace)
```
All of these checks look like they do the same thing: get a list of all pods in the namespace and ensure each of them is running.  Don't we want to select out the collector pod specifically here?

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,257 @@
+		*healthcheck.NewChecker("collector service account exists", "", false, true, time.Time{}, false).
+			WithCheck(func(ctx context.Context) error {
+				// Check for Collector Service Account
+				kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
```
can we build the API client just once in the `jaegerCategory()` instead of in each check?

Pothulapati replied:
Makes sense.

adleong reviewed (commented):
It would also be helpful to see sample output from running this command in the PR description.

Pothulapati commented:
@adleong Addressed the feedback.

Example Output
```bash
tarun in dev in on  kind-kind () linkerd2 on  tarun/jaeger-check [$!?] via 🐹 v1.15.4 took 2s
❯ ./bin/go-run cli jaeger check
linkerd-jaeger
--------------
√ collector service account exists
√ jaeger service account exists
√ collector config map exists
‼ collector pod is running
    collector-69cc44dfbc-8qkxw status is Pending
‼ jaeger pod is running
    jaeger-6f98d5c979-5vnpt status is Pending
√ jaeger extension pods are injected

Status check results are √

tarun in dev in on  kind-kind () linkerd2 on  tarun/jaeger-check [$!?] via 🐹 v1.15.4 took 2s
❯ ./bin/go-run cli jaeger check
linkerd-jaeger
--------------
√ collector service account exists
√ jaeger service account exists
√ collector config map exists
√ collector pod is running
√ jaeger pod is running
√ jaeger extension pods are injected

Status check results are √
```

alpeb commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,232 @@
+
+	checkers := []healthcheck.Checker{}
+	checkers = append(checkers,
+		*healthcheck.NewChecker("collector service account exists", "", false, true, time.Time{}, false).
```
Also I kinda like better how we have things set up in `pkg/healthcheck/healthcheck.go` where you just instantiate the `Checker` struct. Using a long function call like this forces one to go dig into the declaration to see what each argument means.

alpeb replied:
Shouldn't all these checks be `fatal`?

Pothulapati replied:
> how we have things set up in pkg/healthcheck/healthcheck.go where you just instantiate the Checker struct. 

This is not possible here because those fields are not public, so not accessible here. Hence this helper function here. 🤔 To make them public would be a huge change. WDYT?

Pothulapati replied:
> Shouldn't all these checks be fatal?

Currently, They are not and hence I did  the same. I think this was done so that you can get full view on whats happening in the extension. Should they fail instead? 🤔

alpeb replied:
> > how we have things set up in pkg/healthcheck/healthcheck.go where you just instantiate the Checker struct.
> 
> This is not possible here because those fields are not public, so not accessible here. Hence this helper function here.  To make them public would be a huge change. WDYT?

I'm ok making those fields public now that they're used in other packages. The diff certainly will get bigger, but I think it's manageable given it's just string changes.

Pothulapati replied:
Hmm, The Change is manageable, but it could break the API that these structs expose right? to be able to access them anywhere and that creates further expectations on future changes. I remember, there was some discussions around keeping private fields private wherever possible. 

@adleong WDYT?

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+// checkPodsStatus checks if the pod is in running state
```
might want to update this comment to indicate that this checks that all of the pods are in the running state.  And perhaps change the name to describe what it does more specifically.  For example, "checkPodsRunning" or something.

adleong commented on pkg/k8s/api.go:
```diff
@@ -155,6 +155,15 @@ func (kubeAPI *KubernetesAPI) GetPodsByNamespace(ctx context.Context, namespace
 }
 
+// GetPods returns all pods in a given namespace, with the label selector
+func (kubeAPI *KubernetesAPI) GetPods(ctx context.Context, namespace string, selector string) ([]corev1.Pod, error) {
```
Is calling

```
pods, err := kubeAPI.GetPods(ctx, ns, "foo=bar")
```

enough more ergonomic than

```
pods, err := kubeAPI.CoreV1().Pods(ns).List(ctx, metav1.ListOptions{LabelSelector: "foo=bar"})
```

to justify a layer of indirection?  I like the second one better because I can see directly what it does instead of having to look up the GetPods function to see what it does.

Pothulapati replied:
Makes sense. My reasoning was that it's a common function and could be used at multiple places,,

adleong reviewed (commented):
+1 to @alpeb's suggestions

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,228 @@
+	proxyContainer := false
+	for _, containerSpec := range pod.Spec.Containers {
+		if containerSpec.Name == k8s.ProxyContainerName {
+			proxyContainer = true
```
Ah, my motivation for refactoring it this way is that now you really can just `return true` here.

adleong replied:
And then just `return false` at the end of the function if we haven't returned already.

Pothulapati replied:
Yep, Correct! 👍

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,228 @@
+	return nil
+}
+
+func containsProxy(pod *corev1.Pod) bool {
```
nit: if you make this take a `corev1.Pod` instead of a pointer here, you can avoid copying the pod in `checkIfDataPlanePodsExist`.

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,226 @@
+			// Get jaeger Extension Namespace
+			ns, err := getNamespaceOfExtension(jaegerExtensionName)
+			if err != nil {
+				fmt.Fprint(os.Stderr, err.Error())
```
I think this error string needs to have a newline at the end to render properly on the command line.

Pothulapati replied:
Using `fmt.Fprintln` seems to help but this does not seem to be the case everywhere else.

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,226 @@
+			return &ns, err
+		}
+	}
+	return nil, fmt.Errorf("Could not find the namespace for extension %s", name)
```
Let's give this error a call to action:

"Could not find the linkerd-jaeger extension. It can be installed by running `linkerd jaeger install | kubectl apply -f -`\n"

adleong commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,226 @@
+				return nil
+			}))
+
+	checkers = append(checkers,
```
Should we add a retryDeadline to these checks which look for certain pods to be running so that they retry and we get the spinner behavior that the `linkerd check` command does?

Pothulapati replied:
Good Find! :)

adleong reviewed (approved):
Thanks for all the updates, @Pothulapati!  Great work!

kleimkuhler commented on jaeger/cmd/check.go:
```diff
@@ -0,0 +1,226 @@
+		Use:   "check [flags]",
+		Args:  cobra.NoArgs,
+		Short: "Check the Jaeger extension for potential problems",
+		Long: `Check the jaeger extension for potential problems.
```
```suggested
		Long: `Check the Jaeger extension for potential problems.
```

kleimkuhler reviewed (approved):
This looks good! Have some comments about spelling/formatting that should be fixed before merging, but the actual changes look good and work well.
